### PR TITLE
RYA-337 Adding batch queries to MongoDB. Closes #204

### PR DESCRIPTION
**Pull request by amihalik**
## Description

Added a batch query mechanism to MongoDB DAO and simplified MongoDBQueryEngine

### Tests

Ran unit tests 

### Links
[Jira](https://issues.apache.org/jira/browse/RYA-337)

### Checklist
- [ ] Code Review
- [ ] Squash Commits

#### People To Review
@isper3at @pujav65 @meiercaleb

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/386/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.indexing.example' /><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/386/org.apache.rya$rya.indexing.example/testReport'>incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.indexing.example</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/386/org.apache.rya$rya.indexing.example/testReport/(root)/ExamplesTest/MongoRyaDirectExampleTest/'><strong>ExamplesTest.MongoRyaDirectExampleTest</strong></a></li></ul>

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/387/

File: dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java [hunk 1]
```diff
@@ -111,25 +97,20 @@ public MongoDBRdfConfiguration getConf() {
         if (conf == null) {
```
Comment (meiercaleb) [inline dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java]: Check that stmts is not null

File: dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java [hunk 2]
```diff
@@ -140,72 +121,35 @@ public MongoDBRdfConfiguration getConf() {
...
-                // TODO this is gross
-            final RyaStatementCursorIterable iterator = new RyaStatementCursorIterable(
-                    new NonCloseableRyaStatementCursorIterator(new RyaStatementCursorIterator(getCollection(getConf()),
-                            queries, strategy, configuration.getAuthorizations())));
-
-                return iterator;
-            } catch (final Exception e) {
-                throw new RyaDAOException(e);
-            }
+        return query(new BatchRyaQuery(Collections.singleton(ryaQuery.getQuery())));
```
Comment (meiercaleb) [inline dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java]: Null check

File: dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java [hunk 3]
```diff
@@ -140,72 +121,35 @@ public MongoDBRdfConfiguration getConf() {
...
     }
+
     @Override
     public CloseableIterable<RyaStatement> query(final BatchRyaQuery batchRyaQuery)
             throws RyaDAOException {
-         try {
-             final Set<DBObject> queries = new HashSet<DBObject>();
-            for (final RyaStatement statement : batchRyaQuery.getQueries()){
-                queries.add( strategy.getQuery(statement));
+        final Map<RyaStatement, BindingSet> queries = new HashMap<>();
```
Comment (meiercaleb) [inline dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java]: Null check.

File: dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java [hunk 4]
```diff
@@ -86,22 +84,10 @@ public MongoDBRdfConfiguration getConf() {
...
-        final MongoDatabase db = mongoClient.getDatabase(conf.getMongoDBName());
-        final MongoCollection<Document> collection = db.getCollection(conf.getTriplesCollectionName());
-        final RyaStatementCursorIterator iterator = new RyaStatementCursorIterator(collection, queries, strategy,
-                conf.getAuthorizations());
-
-        if (maxResults != null) {
-            iterator.setMaxResults(maxResults);
-        }
-        return iterator;
+        Entry<RyaStatement, BindingSet> entry = new AbstractMap.SimpleEntry<>(stmt, new MapBindingSet());
```
Comment (amihalik) [inline dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java]: done

File: dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java (hunk 1 above)
Comment (amihalik) [inline dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java]: done

File: dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java (hunk 2 above)
Comment (amihalik) [inline dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java]: done

File: dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java (hunk 3 above)
Comment (amihalik) [inline dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/MongoDBQueryEngine.java]: done

File: dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/iter/RyaStatementBindingSetCursorIterator.java [hunk 5]
```diff
@@ -44,20 +48,21 @@
 
 public class RyaStatementBindingSetCursorIterator implements CloseableIteration<Entry<RyaStatement, BindingSet>, RyaDAOException> {
     private static final Logger log = Logger.getLogger(RyaStatementBindingSetCursorIterator.class);
+    
+    private static final int QUERY_BATCH_SIZE = 50;
```
Comment (amihalik) [inline dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/iter/RyaStatementBindingSetCursorIterator.java]: Empirical testing... To support the "show me the first 10" or "first 100" type queries.  I noticed that if we left this at 1000, then the "show me the first 10" or "first 100" type queries were very slow.  However, the "show me all" (where all = 30k), we equally as quick using 50 or 1000.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/410/

**Pull request closed without merging**